[DOCS-1438] Updates to BYOB storage content (#1323)
wandb/docs: content/guides/hosting/data-security/secure-storage-connector.md

- Reports
- Experiment logs
- System metrics
- Console logs

### Buckets

- Experiment files and metrics
- Artifact files
- Media files
- Run files

> - Console logs

Author: ```suggestion
```

- Experiment logs

> - System metrics

Author: ```suggestion
- System metrics
- Console logs
```

Reviewer: Yes look good!

### Buckets

- Experiment files and metrics
- Artifact files
- Media files

> - Run files

Reviewer: I think this change is not needed since we mention console logs specifically below

> - Console logs

Reviewer: These are actually also stored in the database (and uploaded to the bucket as a run file)

Author: Hmm, PTAL at my suggested edits and see if they address this?

> - Exported data in Parquet format

Reviewer: Maybe we can specify here, it's "history metrics and system events in parquet format"

Author: PTAL at the suggestion :)

> - Exported data in Parquet format

Author: ```suggestion
- Exported history metrics and system events in Parquet format
```

Reviewer: Looks good!

| Self-managed | X | X | Instance level BYOB is the default since the instance is fully managed by you. If your self-managed instance is in cloud, you can connect to a cloud-native storage bucket in the same or another cloud for the team-level BYOB. You can also use S3-compatible secure storage like [MinIO](https://github.com/minio/minio) for either of instance or team-level BYOB. |

{{% alert color="secondary" %}}

> Plan your storage bucket layout carefully. After you configure a storage bucket for W&B, you cannot reconfigure it or change its scope, nor can you migrate its data to another bucket and remap relevant references in the main product storage. This applies to storage for Dedicated Cloud and Self-Managed, as well as team-level storage for SaaS Cloud. For questions, contact [support](mailto:support@wandb.com).

Reviewer: Maybe instead of saying that you can't migrate its data, we can say "migrating can be complex" because I know many customers have migrated buckets in the past.

Author: PTAL at the suggested edit to warn off with stronger language.

| Self-managed | X | X | Instance level BYOB is the default since the instance is fully managed by you. If your self-managed instance is in cloud, you can connect to a cloud-native storage bucket in the same or another cloud for the team-level BYOB. You can also use S3-compatible secure storage like [MinIO](https://github.com/minio/minio) for either of instance or team-level BYOB. |

{{% alert color="secondary" %}}

> Plan your storage bucket layout carefully. After you configure a storage bucket for W&B, you cannot reconfigure it or change its scope, nor can you migrate its data to another bucket and remap relevant references in the main product storage. This applies to storage for Dedicated Cloud and Self-Managed, as well as team-level storage for SaaS Cloud. For questions, contact [support](mailto:support@wandb.com).

Author: ```suggestion
Plan your storage bucket layout carefully. After you configure a storage bucket for W&B, migrating its data to another bucket is complex and requires the assistance of W&B. This applies to storage for Dedicated Cloud and Self-Managed, as well as team-level storage for SaaS Cloud. For questions, contact [support](mailto:support@wandb.com).
```

Reviewer: Looks good!

gs://<serviceAccountEmail>:<urlEncodedPrivateKey>@<bucketName>
```

{{% alert %}}

> Because it is hosted in GCP, [SaaS Cloud]({{< relref "/guides/hosting/hosting-options/saas_cloud.md" >}}) does not support connectivity to S3-compatible storage for team-level storage, including cross-cloud connectivity. SaaS Cloud does not have access to the access keys and environment variables used to configure [Dedicated cloud]({{< relref "/guides/hosting/hosting-options/dedicated_cloud.md" >}}) and [Self-managed]({{< relref "/guides/hosting/hosting-options/self-managed.md" >}}) instances.

Reviewer: Is this correct? @levinandrew I thought we supported team-level BYOB for AWS in SaaS

Reviewer 2: yes, we do support team-level BYOB for AWS in SaaS (from our GCP cluster). we just don't support being able to connect using hardcoded access keys, as described in this section https://docs.wandb.ai/guides/hosting/data-security/secure-storage-connector/#cross-cloud-or-s3-compatible-storage-for-team-level-byob

instead we API we are using AssumeRoleWithWebIdentity which uses a JWT

Author: I'm not clear on what needs to be corrected here. I was not intending to change any of the facts in the section, just to adjust the copy.

Reviewer: I think I was confused because I didn't have the full context on the rest of this section. I don't think anything needs to change, just double checking!

Author: Thank you!

    This policy grants your AWS account full access to the key and also assigns the required permissions to the AWS account hosting the W&B Platform. Keep a record of the KMS Key ARN.

1. Provision the S3 Bucket

    Follow these steps to provision the S3 bucket in your AWS account:

    1. Create the S3 bucket with a name of your choice. Optionally create a folder which you can configure as sub-path to store all W&B files.
    1. Enable server side encryption, using the KMS key from the previous step.
    1. Configure CORS with the following policy:

        ```json
        [
          {
              "AllowedHeaders": [
                  "*"
              ],
              "AllowedMethods": [
                  "GET",
                  "HEAD",
                  "PUT"
              ],
              "AllowedOrigins": [
                  "*"
              ],
              "ExposeHeaders": [
                  "ETag"
              ],
              "MaxAgeSeconds": 3000
          }
        ]
        ```

>         {{% alert %}}If data in your bucket expires due to its policy, you can no longer read the history of your runs.{{% /alert %}}

Reviewer: Can we add here that this is specifically in relation to `lifecycle policies` just to be clear what kind of policy this is in reference to.

>         {{% alert %}}If data in your bucket expires due to a lifecycle policy, you can no longer read the history of your runs.{{% /alert %}}

Reviewer: Maybe instead of `you can no longer read the history of your runs`, something like `you may lose the ability to read the history of some runs`.

>         {{% alert %}}If data in your bucket expires due to a lifecycle policy, you can no longer read the history of your runs.{{% /alert %}}

Author: ```suggestion
        {{% alert %}}If data in your bucket expires due to a lifecycle policy, you may lose the ability to read the history of some runs.{{% /alert %}}
```

>         {{% alert %}}If data in your bucket expires due to a lifecycle policy, you may lose the ability to read the history of some runs.{{% /alert %}}

Reviewer 3: Is there any high-level guidance we can provide on this? I know it'll be cloud specific....

Author: Yeah it is Cloud specific. I think this is the sort of thing: https://cloud.google.com/storage/docs/lifecycle, https://learn.microsoft.com/en-us/azure/storage/blobs/lifecycle-management-policy-configure?tabs=azure-portal, https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lifecycle-mgmt.html. Considering all three, I think I will change the text here to "object lifecycle management policy", and link to each cloud's docs in the correct tab.

Author: PTAL

    This policy grants your AWS account full access to the key and also assigns the required permissions to the AWS account hosting the W&B Platform. Keep a record of the KMS Key ARN.

1. Provision the S3 Bucket

    Follow these steps to provision the S3 bucket in your AWS account:

    1. Create the S3 bucket with a name of your choice. Optionally create a folder which you can configure as sub-path to store all W&B files.
    1. Enable server side encryption, using the KMS key from the previous step.
    1. Configure CORS with the following policy:

        ```json
        [
          {
              "AllowedHeaders": [
                  "*"
              ],
              "AllowedMethods": [
                  "GET",
                  "HEAD",
                  "PUT"
              ],
              "AllowedOrigins": [
                  "*"
              ],
              "ExposeHeaders": [
                  "ETag"
              ],
              "MaxAgeSeconds": 3000
          }
        ]
        ```

>         {{% alert %}}If data in your bucket expires due to its policy, you can no longer read the history of your runs.{{% /alert %}}

Author: ```suggestion
        {{% alert %}}If data in your bucket expires due to a lifecycle policy, you can no longer read the history of your runs.{{% /alert %}}
```

    Follow these steps to provision the GCS bucket in your GCP project:

    1. Create the GCS bucket with a name of your choice. Optionally create a folder which you can configure as sub-path to store all W&B files.
    1. Set encryption type to `Google-managed`.
    1. Set the CORS policy with `gsutil`. This is not possible in the UI.

       1. Create a file called `cors-policy.json` locally.
       1. Copy the following CORS policy into the file and save it.

           ```json
           [
             {
               "origin": ["*"],
               "responseHeader": ["Content-Type"],
               "exposeHeaders": ["ETag"],
               "method": ["GET", "HEAD", "PUT"],
               "maxAgeSeconds": 3000
             }
           ]
           ```

>           {{% alert %}}If data in your bucket expires due to a lifecycle policy, you can no longer read the history of your runs.{{% /alert %}}

Author: ```suggestion
          {{% alert %}}If data in your bucket expires due to a lifecycle policy, you may lose the ability to read the history of some runs.{{% /alert %}}
```

      1. Replace `<bucket_name>` with the correct bucket name and run `gsutil`.

          ```bash
          gsutil cors set cors-policy.json gs://<bucket_name>
          ```

      1. Verify the bucket's policy. Replace `<bucket_name>` with the correct bucket name.
        
          ```bash
          gsutil cors get gs://<bucket_name>
          ```

> 1. If you are using [SaaS Cloud]({{< relref "/guides/hosting/hosting-options/saas_cloud.md" >}}) or [Dedicated cloud]({{< relref "/guides/hosting/hosting-options/dedicated_cloud.md" >}}), grant the `storage.admin` role to the GCP service account linked to the W&B Platform. W&B requires this role to check the bucket's CORS configuration and attributes, such as whether object versioning is enabled. If the service account does not have the `storage.admin` role, these checks result in a HTTP 403 error.

Reviewer: This looks good! I was looking for this since I think it was a pretty large source of user confusion before.

    Follow these steps to provision the GCS bucket in your GCP project:

    1. Create the GCS bucket with a name of your choice. Optionally create a folder which you can configure as sub-path to store all W&B files.
    1. Set encryption type to `Google-managed`.
    1. Set the CORS policy with `gsutil`. This is not possible in the UI.

       1. Create a file called `cors-policy.json` locally.
       1. Copy the following CORS policy into the file and save it.

           ```json
           [
             {
               "origin": ["*"],
               "responseHeader": ["Content-Type"],
               "exposeHeaders": ["ETag"],
               "method": ["GET", "HEAD", "PUT"],
               "maxAgeSeconds": 3000
             }
           ]
           ```

>           {{% alert %}}If data in your bucket expires due to its policy, you can no longer read the history of your runs.{{% /alert %}}

Author: ```suggestion
          {{% alert %}}If data in your bucket expires due to a lifecycle policy, you can no longer read the history of your runs.{{% /alert %}}
```

    For the instance level BYOB, if you're not using [this Terraform module](https://github.com/wandb/terraform-azurerm-wandb/tree/main/examples/byob), follow the steps below to provision a Azure Blob Storage bucket in your Azure subscription:

    1. Create a bucket with a name of your choice. Optionally create a folder which you can configure as sub-path to store all W&B files.
    1. Configure the CORS policy on the bucket

        To set the CORS policy through the UI go to the blob storage, scroll down to `Settings/Resource Sharing (CORS)` and then set the following:

        | Parameter | Value |
        | --- | --- |
        | Allowed Origins | `*`  |
        | Allowed Methods | `GET`, `HEAD`, `PUT` |
        | Allowed Headers | `*` |
        | Exposed Headers | `*` |
        | Max Age | `3000` |

>         {{% alert %}}If data in your bucket expires due to a lifecycle policy, you can no longer read the history of your runs.{{% /alert %}}

Author: ```suggestion
        {{% alert %}}If data in your bucket expires due to a lifecycle policy, you may lose the ability to read the history of some runs.{{% /alert %}}
```

    For the instance level BYOB, if you're not using [this Terraform module](https://github.com/wandb/terraform-azurerm-wandb/tree/main/examples/byob), follow the steps below to provision a Azure Blob Storage bucket in your Azure subscription:

    1. Create a bucket with a name of your choice. Optionally create a folder which you can configure as sub-path to store all W&B files.
    1. Configure the CORS policy on the bucket

        To set the CORS policy through the UI go to the blob storage, scroll down to `Settings/Resource Sharing (CORS)` and then set the following:

        | Parameter | Value |
        | --- | --- |
        | Allowed Origins | `*`  |
        | Allowed Methods | `GET`, `HEAD`, `PUT` |
        | Allowed Headers | `*` |
        | Exposed Headers | `*` |
        | Max Age | `3000` |

>         {{% alert %}}If data in your bucket expires due to its policy, you can no longer read the history of your runs.{{% /alert %}}

Author: ```suggestion
        {{% alert %}}If data in your bucket expires due to a lifecycle policy, you can no longer read the history of your runs.{{% /alert %}}
```
